File: SinglePointBatch/path.f90
!Generate a linear path from geometry 1 to geometry 2
!Number of steps along the path is controled by NStep
!Input : internal coordinate definition, geom1, geom2, (optional) NStep
!Output: geom.all
!Dependency: my Fortran-Library, as written in makefile
program main
    use General; use Mathematics; use LinearAlgebra
    use NonlinearOptimization; use Chemistry
    use GeometryTransformation
    implicit none
!Number of steps along the path
    integer::NStep=10
!Molecule information
    integer::NAtoms,intdim,cartdim
    character*2,allocatable,dimension(:)::ElementSymbol
    real*8,allocatable,dimension(:)::ElementNumber,mass,r1,r2,q1,q2
!Work variable
    character*32::chartemp; logical::flag; integer::i,j
    real*8::dbletemp,dbletemp1
    real*8,allocatable,dimension(:)::q,dq,r,rsave
!Initialize
    inquire(file='InternalCoordinateDefinition',exist=flag)
    if(flag) then
        chartemp=''
        call DefineInternalCoordinate(chartemp,intdim)
    else
        chartemp='Columbus7'
        call DefineInternalCoordinate(chartemp,intdim)
    end if
    open(unit=99,file='geom1',status='old')!Read geometry 1
        NAtoms=0; do; read(99,*,iostat=i); if(i/=0) exit; NAtoms=NAtoms+1; end do; rewind 99
        allocate(ElementSymbol(NAtoms)); allocate(ElementNumber(NAtoms))
        allocate(r1(3*NAtoms)); allocate(mass(NAtoms))
        do i=1,NAtoms
            read(99,*)ElementSymbol(i),ElementNumber(i),r1(3*i-2:3*i),mass(i)
        end do
        mass=mass*AMUInAU
    close(99)
    open(unit=99,file='geom2',status='old')!Read geometry 2
        allocate(r2(3*NAtoms))
        do i=1,NAtoms
            read(99,*)chartemp,dbletemp,r2(3*i-2:3*i),dbletemp1
        end do
    close(99)
    inquire(file='NStep',exist=flag)
    if(flag) then
        open(unit=99,file='NStep',status='old'); read(99,*)NStep; close(99)
    end if
    cartdim=3*NAtoms
    allocate(q1(intdim)); q1=InternalCoordinateq(r1,intdim,cartdim)
    allocate(q2(intdim)); q2=InternalCoordinateq(r2,intdim,cartdim)
!Generate a linear path from geometry 1 to geometry 2
    chartemp='assimilate'
    allocate(q(intdim))
    allocate(dq(intdim)); dq=(q2-q1)/dble(NStep+1)
    allocate(r(intdim))
    allocate(rsave(cartdim))
    open(unit=99,file='geom.all',status='replace')
            q=q1+dq
            r=CartesianCoordinater(q,cartdim,intdim,uniquify=chartemp,mass=mass,r0=r1)
            do j=1,NAtoms
                write(99,'(1x,A2,2x,F5.1,4F14.8)')ElementSymbol(j),ElementNumber(j),r(3*j-2:3*j),mass(j)/AMUInAU
            end do
            rsave=r
        do i=2,NStep
            q=q1+dq*dble(i)
            r=CartesianCoordinater(q,cartdim,intdim,uniquify=chartemp,mass=mass,r0=rsave)
            do j=1,NAtoms
                write(99,'(1x,A2,2x,F5.1,4F14.8)')ElementSymbol(j),ElementNumber(j),r(3*j-2:3*j),mass(j)/AMUInAU
            end do
            rsave=r
        end do
    close(99)
end program main

File: SinglePointBatch/path.py
"""
Generate a linear synchronous transit path from geometry 1 to geometry 2
"""

''' Library '''
import argparse
from pathlib import Path
import shutil
import numpy
import FortranLibrary as FL
import basic

''' Routine '''
def parse_args() -> argparse.Namespace: # Command line input
    parser = argparse.ArgumentParser(__doc__)
    parser.add_argument('IntCDefFormat', type=str, help='internal coordinate definition format: Columbus7 or default')
    parser.add_argument('IntCDef', type=Path, help='internal coordinate definition file')
    parser.add_argument('geom1',   type=Path, help='Columbus7 geometry file 1')
    parser.add_argument('geom2',   type=Path, help='Columbus7 geometry file 2')
    parser.add_argument('-n','--NSteps', type=int, default=10, help='number of linear synchronous transit steps (default = 10)')
    parser.add_argument('-o','--output', type=Path, default='geom.data', help='output file (default = geom.data), will append if already exists')
    args = parser.parse_args()
    return args

if __name__ == "__main__":
    ''' Initialize '''
    args = parse_args()
    # Define internal coordinate
    if args.IntCDefFormat == 'Columbus7': IntCDef = Path('intcfl') 
    else: IntCDef = Path('InternalCoordinateDefinition')
    if args.IntCDef != IntCDef:
        shutil.copy(args.IntCDef, IntCDef)
        intdim = FL.DefineInternalCoordinate(args.IntCDefFormat)
        IntCDef.unlink()
    else:
        intdim = FL.DefineInternalCoordinate(args.IntCDefFormat)
    # Read geometry
    NAtoms, symbol, number, r1, mass = basic.read_geom(args.geom1)
    NAtoms, symbol, number, r2, mass = basic.read_geom(args.geom2)
    cartdim = 3 * NAtoms
    q1 = numpy.empty(intdim); FL.InternalCoordinateq(r1, q1, cartdim, intdim)
    q2 = numpy.empty(intdim); FL.InternalCoordinateq(r2, q2, cartdim, intdim)
    ''' Do the job '''
    dq = numpy.empty(intdim); dq[:] = (q2[:]-q1[:]) / (args.NSteps+1)
    q = numpy.empty(intdim)
    r = numpy.empty(cartdim); rsave = r1.copy()
    for i in range(1, args.NSteps+1):
        q[:] = q1[:] + float(i) * dq[:]
        FL.CartesianCoordinater(q, r, intdim, cartdim, uniquify='assimilate', mass=mass, r0=rsave)
        basic.write_geom(args.output, NAtoms, symbol, number, r, mass)
        rsave[:] = r[:]